File: leetcode/2020/[Medium] binary-tree-preorder-traversal/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        traverse(list, root);
        return list;
    }
    
    private void traverse(List<Integer> list, TreeNode node) {
        if (node == null) return;
        list.add(node.val);
        traverse(list, node.left);
        traverse(list, node.right);
    }
}

File: leetcode/2020/[Medium] binary-tree-preorder-traversal/Solution.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(var _value: Int) {
 *   var value: Int = _value
 *   var left: TreeNode = null
 *   var right: TreeNode = null
 * }
 */
object Solution {
    def preorderTraversal(root: TreeNode): List[Int] = root match {
        case null => Nil
        case _ => root.value :: preorderTraversal(root.left) ::: preorderTraversal(root.right)
    }
}

